applications/StructuralMechanicsApplication/custom_utilities/large_displacement_variables.cpp

```diff
@@ -0,0 +1,328 @@
+    KRATOS_CATCH("");
+}
+
+LargeDisplacementKinematicVariables::LargeDisplacementKinematicVariables(
+    LargeDisplacementDifferentialVariables& rDiffVars, bool IsAxisymmetric)
+    : mrDiffVars(rDiffVars), mIsAxisymmetric(IsAxisymmetric)
+{
+}
+
+const Matrix& LargeDisplacementKinematicVariables::StrainTensor(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mCurrentIntegrationIndex)
+        Recalculate(IntegrationIndex);
+    return mStrainTensor;
+    KRATOS_CATCH("");
+}
+
+Vector& LargeDisplacementKinematicVariables::StrainVector(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mStrainVectorIntegrationIndex)
+    {
+        mStrainVector =
+            MathUtils<double>::StrainTensorToVector(StrainTensor(IntegrationIndex));
+        mStrainVectorIntegrationIndex = IntegrationIndex;
+    }
+    return mStrainVector;
+    KRATOS_CATCH("");
+}
+
+const Matrix& LargeDisplacementKinematicVariables::B(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    if (IntegrationIndex != mCurrentIntegrationIndex)
+        Recalculate(IntegrationIndex);
+    return mB;
+    KRATOS_CATCH("");
+}
+
+void LargeDisplacementKinematicVariables::Recalculate(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    CalculateStrainTensor(IntegrationIndex);
+    CalculateB(IntegrationIndex);
+    mCurrentIntegrationIndex = IntegrationIndex;
+    KRATOS_CATCH("");
+}
+
+void LargeDisplacementKinematicVariables::CalculateStrainTensor(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    const Matrix& rF = mrDiffVars.F(IntegrationIndex, mIsAxisymmetric);
+    MatrixVectorUtils::InitializeMatrix(mStrainTensor, rF.size2(), rF.size2(), false);
+    noalias(mStrainTensor) = 0.5 * prod(trans(rF), rF);
+    for (std::size_t i = 0; i < mStrainTensor.size1(); ++i)
+        mStrainTensor(i, i) -= 0.5;
+    KRATOS_CATCH("");
+}
+
+void LargeDisplacementKinematicVariables::CalculateB(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    if (mIsAxisymmetric)
+        CalculateB_Axisymmetric(IntegrationIndex);
+    else if (mrDiffVars.GetGeometry().WorkingSpaceDimension() == 2)
+        CalculateB_2D(mrDiffVars.F(IntegrationIndex),
+                      mrDiffVars.DN_DX0(IntegrationIndex), mB);
+    else
+        CalculateB_3D(mrDiffVars.F(IntegrationIndex),
+                      mrDiffVars.DN_DX0(IntegrationIndex), mB);
+    KRATOS_CATCH("");
+}
+
+void LargeDisplacementKinematicVariables::CalculateB_Axisymmetric(std::size_t IntegrationIndex)
+{
+    KRATOS_TRY;
+    const auto& r_geom = mrDiffVars.GetGeometry();
+    const unsigned int number_of_nodes = r_geom.PointsNumber();
+    const unsigned int dimension = r_geom.WorkingSpaceDimension();
+    const unsigned int strain_size = 4;
+    const Matrix& rF = mrDiffVars.F(IntegrationIndex, true);
+    const Matrix& rDN_DX0 = mrDiffVars.DN_DX0(IntegrationIndex);
+    Vector N = row(r_geom.ShapeFunctionsValues(), IntegrationIndex);
```

Comment:
are u intentionally makin a copy?

Reply:
I didn't think a reference was possible here. Now I tried it and it compiles fine so I will change it.

Reply:
done.